File: Code/Iterator.dotNet/Iterator.dotNet/Iterator.dotNet/Iterators/DepthFirstIterator.cs
using Iterator.DataStructure;
using Iterator.Model;

namespace Iterator.Iterators
{
    /// <summary>
    /// Iterate over a IIterableCollection with DepthFirst
    /// </summary>
    internal class DepthFirstIterator : IIterator
    {
        private Queue<TreeNode> _curr = null;

        private void InsertElementInQueue(TreeNode e)
        {
            if (e == null) return;

            this._curr.Enqueue(e);

            if (e.Left != null) InsertElementInQueue(e.Left);

            if (e.Right != null) InsertElementInQueue(e.Right);
        }

        public DepthFirstIterator(TreeCollection treeCollection)
        {
            _curr = new Queue<TreeNode>();
            InsertElementInQueue(treeCollection.GetRoot());

        }

        /* As IIterator */

        public Item GetNext()
        {
            var ret = _curr.Dequeue();
            return ret.Item;
        }

        public bool HasMore() => this._curr.Count > 0;
    }
}

File: Code/Iterator.Java/src/iterators/DepthFirstIterator.java
package iterators;

import java.util.LinkedList;
import java.util.Queue;

import datastructure.TreeNode;
import model.Item;
import datastructure.TreeCollection;

/**
 * Iterate over a IIterableCollection with DepthFirst
 */
public class DepthFirstIterator implements IIterator {

	private Queue<TreeNode> _curr = null;
	
	/**
	 * Insert elements in the right order
	 * @param e
	 */
	private void insertElementInQueue(TreeNode e) {
		if(e == null) return;
		
		this._curr.add(e);
		
		if(e.getLeft() != null) {
			insertElementInQueue(e.getLeft());
		}
		
		if(e.getRight() != null) { 
			insertElementInQueue(e.getRight());		
		}
	}
	
	/* Here, we don't care of passing a concrete object instead of its
	 * interface, i.e., IITerableCollection. Why?
	 */
	public DepthFirstIterator(TreeCollection treeCollection) {
		this._curr = new LinkedList<TreeNode>();
		insertElementInQueue(treeCollection.getRoot());
	}

	@Override
	public Item getNext() {
		TreeNode ret = this._curr.element();
		this._curr.remove(ret);
		return ret.getItem();
	}

	@Override
	public boolean hasMore() {
		return !this._curr.isEmpty();
	}

}
